### PR TITLE
Fix(Counter & Lists): Add validation for `priorities`

### LINKED ISSUE #3684: No validation on GameServerAllocation.Priorities
**What happened**:

You can pass in invalid values for priority.type and priority.order on a `GameServerAllocation`

**What you expected to happen**:

Return an error stating the values are invalid (this happens with a Fleet - but it has the OpenAPI Spec).

**How to reproduce it (as minimally and precisely as possible)**:

Create this Fleet:

```yaml
apiVersion: agones.dev/v1
kind: Fleet
metadata:
  name: simple-game-server
spec:
  replicas: 2
  template:
    spec:
      ports:
        - name: default
          containerPort: 7654
      counters:
        rooms:
          count: 0
          capacity: 10
      lists:
        players:
          values: []
      template:
        spec:
          containers:
            - name: simple-game-server
              image: us-docker.pkg.dev/agones-images/examples/simple-game-server:0.27
              resources:
                requests:
                  memory: 64Mi
                  cpu: 20m
                limits:
                  memory: 64Mi
                  cpu: 20m
```
Then `kubectl create -f <this file>`:

```yaml
apiVersion: allocation.agones.dev/v1
kind: GameServerAllocation
spec:
  scheduling: Packed
  priorities:
    - type: John
      key: rooms
      order: Dog
  selectors:
    # First check to see if we can back-fill an already allocated `GameServer`
    - gameServerState: Allocated
      matchLabels:
        agones.dev/fleet: simple-game-server
      counters:
        rooms:
          minAvailable: 1
   # If we can't, then go get a `Ready` `GameServer`.
    - gameServerState: Ready
      matchLabels:
        agones.dev/fleet: simple-game-server
      counters:
        rooms:
          minAvailable: 1
  counters:
    rooms:
      action: Increment
      amount: 1
```

It will complete with no error.

**Anything else we need to know?**:

Since GameServerAllocation doesn't have the OpenAPI spec, we will have to do this somewhere in the `.Validate()` function of GameServerAllocation.

https://github.com/googleforgames/agones/blob/1c536c0aec680e8b013479bc15481c4ae47e3185/pkg/apis/allocation/v1/gameserverallocation.go#L376

**Environment**:
- Agones version: dev
- Kubernetes version (use `kubectl version`): 1.28
- Cloud provider or hardware configuration: GKE
- Install method (yaml/helm): helm
- Troubleshooting guide log(s): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lacroixthomas**
**What type of PR is this?**

/kind bug

**What this PR does / Why we need it**:
Add validation of the priorities for the Counts and Lists feature.

**Which issue(s) this PR fixes**:

Closes #3684 

**Special notes for your reviewer**:
This is my first contribution on this project, I'm open to any suggestions / changes, thanks !

### CODE REVIEW

**Comment by markmandel** [comment]
Thanks for the submission! A lot of us are at GDC this week, so we might not get to review this until next week.

File: pkg/apis/allocation/v1/gameserverallocation.go [hunk 1]
```diff
@@ -465,6 +465,25 @@ func validateLists(lists map[string]ListSelector, fldPath *field.Path) field.Err
...
+	for index, priority := range priorities {
+		keyPath := fldPath.Index(index)
+		if priority.Type != agonesv1.GameServerPriorityCounter && priority.Type != agonesv1.GameServerPriorityList {
+			allErrs = append(allErrs, field.Invalid(keyPath, priority.Type, "type must be \"Counter\" or \"List\""))
+		}
+		if priority.Key == "" {
+			allErrs = append(allErrs, field.Invalid(keyPath, priority.Type, "key must not be nil"))
+		}
+		if priority.Order != agonesv1.GameServerPriorityAscending && priority.Order != agonesv1.GameServerPriorityDescending {
+			allErrs = append(allErrs, field.Invalid(keyPath, priority.Order, "order must be \"Increment\" or \"Decrement\""))
```
Comment (igooch) [inline pkg/apis/allocation/v1/gameserverallocation.go:480]:
```suggestion
			allErrs = append(allErrs, field.Invalid(keyPath, priority.Order, "order must be \"Ascending\" or \"Descending\""))
```

**Review by igooch — CHANGES_REQUESTED** [review]
Thanks for the fix! One small comment, otherwise LGTM.

File: pkg/apis/allocation/v1/gameserverallocation.go (hunk 1 above)
Comment (lacroixthomas) [reply pkg/apis/allocation/v1/gameserverallocation.go]: Youps missed that, just pushed the changes, Thanks !

**Review by igooch — APPROVED** [review]
LGTM

**Pull request merged**